File: scripts/algorithms/M/Minimum Cost to Cut a Stick/Minimum Cost to Cut a Stick.cpp
// Runtime: 233 ms (Top 8.43%) | Memory: 10.3 MB (Top 39.50%)
class Solution {
public:
    long cutMin(int i, int j, vector<int>&c, vector<vector<int>>&dp)
    {
        if(i>j) return 0;
        if(dp[i][j]!=-1) return dp[i][j];
        long mini = INT_MAX;
        for(int ind=i;ind<=j;ind++)
        {
            long cost = c[j+1]-c[i-1]+cutMin(i,ind-1,c,dp)+cutMin(ind+1,j,c,dp);
            mini = min(mini,cost);
        }
        return dp[i][j] = mini;
    }
    int minCost(int n, vector<int>& cuts) {
        int c = cuts.size();
        cuts.push_back(0);
        cuts.push_back(n);
        sort(cuts.begin(),cuts.end());
        vector<vector<int>>dp(c+1,vector<int>(c+1,-1));
        return cutMin(1,c,cuts,dp);
    }
};

File: scripts/algorithms/M/Minimum Cost to Cut a Stick/Minimum Cost to Cut a Stick.rs
// Runtime: 4 ms (Top 83.33%) | Memory: 2.20 MB (Top 100.0%)

impl Solution {
    pub fn min_cost(n: i32, cuts: Vec<i32>) -> i32 {
        let mut a = Vec::with_capacity(cuts.len() + 2);
        a.push(0);
        a.push(n);
        a.extend(cuts);
        a.sort_unstable();

        let mut dp = vec![vec![0; a.len()]; a.len()];

        for len in 2..a.len() {
            for i in 0..a.len() - len {
                let j = i + len;

                let mut t = i32::MAX;
                for k in i + 1..j {
                    t = t.min(dp[i][k] + dp[k][j]);
                }
                dp[i][j] = t + a[j] - a[i];
            }
        }

        dp[0][a.len() - 1]
    }
}
